### PR TITLE
Migrate mailing lists to their own API endpoints

### PR DESCRIPTION
**Pull request by jchristgit**
Add a new model for the bot to store its mailing list state in, as opposed to the current JSON blob in the BotSetting table. Migrate the existing settings from the BotSetting table into the new model.

### CODE REVIEW

File: pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py
Comment (shtlrs) [inline pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py]: Wouldn't it be better to squash the schema migrations into one ?

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (shtlrs) [inline pydis_site/apps/api/viewsets/bot/mailing_list.py:28]: Is there a reason why you're returning a list of objects instead of a list of the hashes directly ?

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (shtlrs) [inline pydis_site/apps/api/viewsets/bot/mailing_list.py]: We can void the unnecessary query by doing the `isinstance` check before fetching the object

**Review by shtlrs — CHANGES_REQUESTED** [review]
Great changes ! 
This is just a code review for the timebeing, will test in a later instance.

File: pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py
Comment (jchristgit) [reply pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py]: I've squashed the final two, however for the previous migrations I've intentionally opted to keep them separate because one is an elidable data migration. I suggest once this and the alternate account PR is merged we make a "big squash"

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (jchristgit) [reply pydis_site/apps/api/viewsets/bot/mailing_list.py:28]: Yes. I was lazy :smiley:

I've changed and pushed this.

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (jchristgit) [reply pydis_site/apps/api/viewsets/bot/mailing_list.py]: Good idea, changed.

File: pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py
Comment (jchristgit) [reply pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py]: Hello. My lawyer has advised me to squash the migrations as you requested. This is now live.

Comment (shtlrs) [reply pydis_site/apps/api/migrations/0096_mailinglistseenitem_unique_list_and_hash.py]: Thanks !
Great lawyer btw, you should give him a raise.

File: pydis_site/apps/api/migrations/0094_migrate_mailing_listdata.py
Comment (shtlrs) [inline pydis_site/apps/api/migrations/0094_migrate_mailing_listdata.py]: The `reverse_code` param defaults to None, so no need to have the `do_nothing`  method def.

**Review by shtlrs — APPROVED** [review]
Works great, thanks !
I've marked this as do not merge untiil the bot pr is ready.

File: pydis_site/apps/api/migrations/0094_migrate_mailing_listdata.py
Comment (jchristgit) [reply pydis_site/apps/api/migrations/0094_migrate_mailing_listdata.py]: Oh, I actually left this in by accident. I had to add this to allow me to roll back the migration properly (if you don't specify the reverse, it will error out as being irreversible). I've removed it now for The Big Squash 2023

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (shtlrs) [inline pydis_site/apps/api/viewsets/bot/mailing_list.py]: As I was making the bot side, I realised that the fact that we can't create mailing lists is an issue for the following reasons:
1. We don't have any migrations that create theses mailing lists, so if you delete a container that used to have th old table structure, you'll never end up with the target mailing lists.
2. We used to workaround this bot side thanks to the `sync_mailing_lists` method that creates the proper JSON that will be dumped in the db, this is no longer true
3. The bot will keep on getting `404` as it has as hardcoded mailing list that it tracks, so when it tries to post to site, it will complain saying it can't find such registered list.


We should use `ListCreateModelMixin` instead to allow the bot to create the mailing lists.

**Review by shtlrs — CHANGES_REQUESTED** [review]

File: pydis_site/apps/api/viewsets/bot/mailing_list.py
Comment (jchristgit) [reply pydis_site/apps/api/viewsets/bot/mailing_list.py]: Do you mean `CreateModelMixin`? I can't find anything about `ListCreateModelMixin`.

Comment (shtlrs) [reply pydis_site/apps/api/viewsets/bot/mailing_list.py]: I always get confused, I was talking about the ApiView.

Yes you're right, it's the CreateModelMixin

Comment (jchristgit) [reply pydis_site/apps/api/viewsets/bot/mailing_list.py]: Pushed this.

**Review by shtlrs — APPROVED** [review]
On The House

**Comment by Xithrius** [comment]
Are we good to go on removing the do not merge label?

**Review by Xithrius — APPROVED** [review]

**Pull request merged**
